README.md
# Tink Angular format helper

v1.2.4

## What is this repository for?

The Tink Angular format helper helps with formats.

Tink is an in-house developed easy-to-use front end framework for quick prototyping and simple deployment of all kinds of websites and apps, keeping a uniform and consistent look and feel.

## Setup

### Prerequisites

* nodeJS [http://nodejs.org/download/](http://nodejs.org/download/)
* bower: `npm install -g bower`

### Install

1. Go to the root of your project and type the following command in your terminal:
   `bower install tink-helper-format-angular --save`

2. Add the following files to your project:
  `bower_components/randexp/build/randexp.min.js`

  `bower_components/tink-helper-format-angular/dist/tink-helper-format-angular.js`

  `bower_components/tink-helper-date-angular/dist/tink-helper-date-angular.js`

  `bower_components/tink-helper-safe-apply-angular/dist/tink-helper-safe-apply-angular.js`

3. On http://tink.digipolis.be you will find all necessary documentation.

## Contribution guidelines

* If you're not sure, drop us a note
* Fork this repo
* Do your thing
* Create a pull request

## Who do I talk to?

* Jasper Van Proeyen - jasper.vanproeyen@digipolis.be - Lead front-end
* Tom Wuyts - tom.wuyts@digipolis.be - Lead UX
* [The hand](https://www.youtube.com/watch?v=_O-QqC9yM28)


CHANGELOG.md
# Changelog Tink Angular format helper

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

<!--
## [Unreleased] - [unreleased]

### Added
### Changed
### Deprecated
### Removed
### Fixed
### Security
-->

## [1.2.4] - 2016-01-15

### Fixed
- Fixed a console error being triggered only in Internet Explorer



## [1.2.3] - 2016-01-04

### Removed
- Removed console logs



## [1.2.2] - 2015-12-02

### Removed
- Removed randexp from build



## [1.2.1] - 2015-12-02

### Added
- Added randexp to Bower

### Fixed
- Fixed a bug in the 'contenteditable' field (only for Webkit browsers)



## [1.2.0] - 2015-12-02

### Added
- Added regex validation



## [1.1.6] - 2015-11-30

### Fixed
- Fixed issue with ngModel watch



## [1.1.5] - 2015-11-18

### Fixed
- Fixed issue with disabling elements and tabindex



## [1.1.4] - 2015-09-28

### Fixed
- Fixed change event that wasn't firing



## [1.1.3] - 2015-07-22

### Changed
- Upgraded angular to version 1.4.3

### Fixed
- Fixed firing 2 watches because of viewValue.



## [1.1.2] - 2015-05-29

### Fixed
- Fixed when disabled, also the user interaction is disabled.



## [1.1.1] - 2015-05-28

### Fixed
- Quick fix for model not becoming empty when a wrong format is inputted



## [1.1.0] - 2015-05-20

### Added
- value stays when wrong format

### Fixed
- data model becomes null when delete everything



## [1.0.0] - 2015-04-22

Initial release


test/karma.conf.js
// Karma configuration
// http://karma-runner.github.io/0.12/config/configuration-file.html
// Generated on 2014-12-01 using
// generator-karma 0.8.3

module.exports = function(config) {
  'use strict';

  config.set({
    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,

    // base path, that will be used to resolve files and exclude
    basePath: '../',

    // testing framework to use (jasmine/mocha/qunit/...)
    frameworks: ['jasmine'],

    // list of files / patterns to load in the browser
    files: [
      'bower_components/jquery/dist/jquery.js',
      'bower_components/angular/angular.js',
      'bower_components/angular-mocks/angular-mocks.js',
      'bower_components/angular-animate/angular-animate.js',
      'bower_components/angular-cookies/angular-cookies.js',
      'bower_components/angular-resource/angular-resource.js',
      'bower_components/angular-route/angular-route.js',
      'bower_components/angular-sanitize/angular-sanitize.js',
      'bower_components/angular-touch/angular-touch.js',
      'bower_components/tink-helper-date-angular/dist/tink-helper-date-angular.js',
      'bower_components/tink-helper-safe-apply-angular/dist/tink-helper-safe-apply-angular.js',
      'bower_components/tink-iban-angular/dist/tink-iban-angular.js',
      'bower_components/randexp/build/randexp.min.js',
      'src/scripts/{,*/}*.js',
      'src/{,*/}',
      'test/mock/{,*/}*.js',
      'test/spec/{,*/}*.js',
      'src/templates/{,*/}*.html'
    ],

    // list of files / patterns to exclude
    exclude: [],

    // web server port
    port: 8080,

    // Start these browsers, currently available:
    // - Chrome
    // - ChromeCanary
    // - Firefox
    // - Opera
    // - Safari (only Mac)
    // - PhantomJS
    // - IE (only Windows)
    browsers: [
      'PhantomJS'
    ],
    preprocessors:{
      'src/templates/{,*/}*.html': 'html2js'
    },
    ngHtml2JsPreprocessor: {
      stripPrefix: '../templates/',  // <-- change as needed for the project
      // include beforeEach(module('templates')) in unit tests
      moduleName: 'templates'
    },

    // Which plugins to enable
    plugins: [
      'karma-jasmine',
      'karma-phantomjs-launcher',
      'karma-chrome-launcher',
      'karma-sauce-launcher',
      'karma-coverage',
      'karma-ng-html2js-preprocessor'
    ],

    // Continuous Integration mode
    // if true, it capture browsers, run tests and exit
    singleRun: false,

    colors: true,

    // level of logging
    // possible values: LOG_DISABLE || LOG_ERROR || LOG_WARN || LOG_INFO || LOG_DEBUG
    logLevel: config.LOG_INFO,

    // Uncomment the following lines if you are using grunt's server to run the tests
    // proxies: {
    //   '/': 'http://localhost:9000/'
    // },
    // URL root prevent conflicts with the site root
    // urlRoot: '_karma_'
  });
};


test/spec/tink-helper-format-angular.js
'use strict';
describe('tink-helper-format-angular', function() {

  var bodyEl = $('body'), sandboxEl;
  var $compile, $templateCache, scope;

  beforeEach(module('tink.formathelper'));

  beforeEach(inject(function (_$rootScope_, _$compile_, _$templateCache_) {
    scope = _$rootScope_.$new();
    $compile = _$compile_;
    $templateCache = _$templateCache_;
    bodyEl.html('');
    sandboxEl = $('<div>').attr('id', 'sandbox').appendTo(bodyEl);
  }));

  afterEach(function() {
    scope.$destroy();
    sandboxEl.remove();
  });

  function compileDirective(template, locals) {
    template = templates[template];
    angular.extend(scope, angular.copy(template.scope || templates['default'].scope), locals);
    var element = $(template.element).appendTo(sandboxEl);
    element = $compile(element)(scope);
    scope.$digest();
    return jQuery(element[0]);
  }

  function triggerInput(el,key){
    $(el).focus();
    for(var i =0; i< key.length;i++){
      var e = jQuery.Event('keypress');
      e.which = key.charCodeAt(i);
      $(el).trigger(e);
    }
  }

  var templates = {
    'default': {
      scope: {data:null},
      element: '<div tink-format-input ng-model="data"></div>'
    },
    'required': {
      scope: {data:null},
      element: '<div tink-format-input required ng-model="data"></div>'
    },
    'dates': {
      scope: {data:null},
      element: '<div tink-format-input min-date="minDate" max-date="maxDate" ng-model="data"></div>'
    }
  };


  describe('default', function() {
    it('should run this basic setup',function(){
      var element = compileDirective('default');
      var scope = element.isolateScope();
      scope.ctrl.setValue('20/01/2010');
      expect(scope.ctrl.getValue()).toBe('20/01/2010');
    });

    it('should have the right html',function(){
      var element = compileDirective('default');
      var scope = element.isolateScope();
      scope.ctrl.setValue('20/01/2010');
      expect(scope.ctrl.element.html()).toBe('20<span class="placeholder">/</span>01<span class="placeholder">/</span>2010');
    });

    it('should be valid without required',function(){
      var element = compileDirective('default');
      var scope = element.isolateScope();
      expect(scope.ctrl.ngControl.$valid).toBe(true);
      expect(scope.ctrl.ngControl.$invalid).toBe(false);
    });

    it('should be invalid with required attribute',function(){
      var element = compileDirective('required');
      var scope = element.isolateScope();
      expect(scope.ctrl.ngControl.$valid).toBe(false);
      expect(scope.ctrl.ngControl.$invalid).toBe(true);
      expect(scope.ctrl.ngControl.$error.required).toBe(true);
    });

    it('should be invalid with wrong date',function(){
      var element = compileDirective('required');
      var scope = element.isolateScope();
      triggerInput(element.find('div.faux-input'),'20-01-199');
      element.find('div.faux-input').blur();
      scope.$digest();
      expect(scope.ctrl.ngControl.$error.date).toBe(true);
    });

    it('should be valid when proper date',function(){
      var element = compileDirective('required');
      var scope = element.isolateScope();
      triggerInput(element.find('div.faux-input'),'20-01-1992');
      element.find('div.faux-input').blur();
      scope.$digest();
      expect(scope.ctrl.ngControl.$valid).toBe(true);
      expect(scope.ctrl.ngControl.$invalid).toBe(false);
    });

    it('should be valid when proper date',function(){
      var element = compileDirective('required');
      var scope = element.isolateScope();
      triggerInput(element.find('div.faux-input'),'20-01-1992');
      element.find('div.faux-input').blur();
      scope.$digest();
      expect(scope.ctrl.ngControl.$valid).toBe(true);
      expect(scope.ctrl.ngControl.$invalid).toBe(false);
    });
    
    it('should be invalid when min date is wrong',function(){
      var element = compileDirective('dates',{minDate:new Date()});
      var scope = element.isolateScope();
      triggerInput(element.find('div.faux-input'),'20-01-1992');
      element.find('div.faux-input').blur();
      scope.$digest();
      expect(scope.ctrl.ngControl.$valid).toBe(false);
      expect(scope.ctrl.ngControl.$invalid).toBe(true);
      expect(scope.ctrl.ngControl.$error['date-min']).toBe(true);
    });

    it('should be invalid when max date is wrong',function(){
      var element = compileDirective('dates',{maxDate:new Date(1992,0,20)});
      var scope = element.isolateScope();
      triggerInput(element.find('div.faux-input'),'21-01-1992');
      element.find('div.faux-input').blur();
      scope.$digest();
      expect(scope.ctrl.ngControl.$valid).toBe(false);
      expect(scope.ctrl.ngControl.$invalid).toBe(true);
      expect(scope.ctrl.ngControl.$error['date-max']).toBe(true);
    });    
  });
});

Gruntfile.js
'use strict';

module.exports = function (grunt) {

  // Load grunt tasks automatically
  require('load-grunt-tasks')(grunt);

  // Time how long tasks take. Can help when optimizing build times
  require('time-grunt')(grunt);

  // Check for a new version
  var bump = grunt.option('bump') || '';
  var currentversion = require('./package.json').version;
  if (bump !== '') {
    var semversion = require('semver');
    currentversion = semversion.inc(currentversion, bump);
  }

  // Configurable paths for the application
  var appConfig = {
    app: require('./bower.json').appPath || 'src',
    dist: 'dist',
    version: currentversion,
    module: require('./package.json').name,
    domain: 'tink.formatdate'
  };

  // Define the configuration for all the tasks
  grunt.initConfig({

    // Project settings
    yeoman: appConfig,
    watch: {
      bower: {
        files: ['bower.json'],
        tasks: ['wiredep']
      },
      js: {
        files: ['<%= yeoman.app %>/scripts/{,*/}*.js'],
        tasks: ['newer:jshint:all'],
        options: {
          livereload: '<%= connect.options.livereload %>'
        }
      },
      jsTest: {
        files: ['test/spec/{,*/}*.js'],
        tasks: ['newer:jshint:test', 'karma']
      },
      sass: {
        // files: ['<%= yeoman.app %>/styles/{,*/}*.{scss,sass}'],
        files: ['<%= yeoman.app %>/styles/{,*/}*.{scss,sass}'],
        tasks: ['sass:server', 'autoprefixer']
      },
      html: {
        files: ['<%= yeoman.app %>/templates/{,*/}*.{htm,html}'],
        tasks: ['ngtemplates']
      },
      gruntfile: {
        files: ['Gruntfile.js']
      },
      livereload: {
        options: {
          livereload: '<%= connect.options.livereload %>'
        },
        files: [
          '<%= yeoman.app %>/index.html',
          '<%= yeoman.app %>/{,*/}*.html',
          '.tmp/**/*.css',
          '<%= yeoman.app %>/images/{,*/}*.{png,jpg,jpeg,gif,webp,svg}'
        ]
      }
    },
    ngtemplates:  {
      app:        {
        options: {
          module: '<%= yeoman.domain %>',
          standalone:false,
          htmlmin: {
            collapseBooleanAttributes: true,
            collapseWhitespace: true,
            removeAttributeQuotes: true,
            removeComments: true, // Only if you don't use comment directives!
            removeEmptyAttributes: true,
            removeRedundantAttributes: true,
            removeScriptTypeAttributes: true,
            removeStyleLinkTypeAttributes: true,
            conservativeCollapse: true,
            preserveLineBreaks: true
          }
        },
        cwd: 'src',
        src: 'templates/**.html',
        dest: '<%= yeoman.dist %>/scripts/templates.js'
      }
    },
    bump: {
      options: {
        files: ['package.json', 'bower.json'],
        updateConfigs: [],
        commit: true,
        commitMessage: 'Release v%VERSION%',
        commitFiles: ['-a'],
        createTag: true,
        tagName: 'v%VERSION%',
        tagMessage: 'v%VERSION%',
        push: true,
        pushTo: 'origin',
        gitDescribeOptions: '--tags --always --abbrev=1 --dirty=-d',
        globalReplace: false,
        prereleaseName: false,
        regExp: false
      }
    },
    jshint: {
      options: {
        jshintrc: '.jshintrc',
        reporter: require('jshint-stylish')
      },
      all: {
        src: [
          'Gruntfile.js',
          '<%= yeoman.app %>/scripts/{,*/}*.js'
        ]
      },
     test: {
        options: {
          jshintrc: 'test/.jshintrc'
        },
        src: ['test/spec/{,*/}*.js']
      }
    },
    connect: {
      options: {
        port: 9000,
        // Change this to '0.0.0.0' to access the server from outside.
        hostname: '0.0.0.0',
        livereload: 35729
      },
      livereload: {
        options: {
          open: true,
          middleware: function (connect) {
            return [
              connect.static('.tmp'),
              connect().use(
                '/bower_components',
                connect.static('./bower_components')
              ),
              connect.static(appConfig.app)
            ];
          }
        }
      },
      test: {
        options: {
          port: 9001,
          middleware: function (connect) {
            return [
              connect.static('.tmp'),
              connect.static('test'),
              connect().use(
                '/bower_components',
                connect.static('./bower_components')
              ),
              connect.static(appConfig.app)
            ];
          }
        }
      },
      dist: {
        options: {
          open: true,
          base: '<%= yeoman.dist %>'
        }
      }
    },
    clean: {
      dist: {
        files: [{
          dot: true,
          src: [
            '.tmp',
            '<%= yeoman.dist %>/{,*/}*',
            '!<%= yeoman.dist %>/.git*'
          ]
        }]
      },
      server: {
        files: [{
          src: [
            '.tmp',
            '<%= yeoman.dist %>/scripts'
          ]
        }]
      }
    },
    autoprefixer: {
      options: {
        browsers: ['last 3 versions']
      },
      dist: {
        files: [{
          expand: true,
          cwd: '.tmp/styles',
          src: '{,*/}*.css',
          dest: '.tmp/styles'
        }]
      }
    },
    wiredep: {
      app: {
        src: ['<%= yeoman.app %>/index.html'],
        ignorePath: /\.\.\//
      },
      sass: {
        src: ['<%= yeoman.app %>/styles/{,*/}*.{scss,sass}'],
        ignorePath: /(\.\.\/){1,2}bower_components\//
      }
    },
    sass: {
      options: {
        imagePath:'../images',
        includePaths: [
            'bower_components'
        ]
      },
      dist: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.app %>/styles',
          src: ['*.scss'],
          dest: '.tmp/styles',
          ext: '.css'
        }]
      },
      server: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.app %>/styles',
          src: ['*.scss'],
          dest: '.tmp/styles',
          ext: '.css'
        }]
      }
    },
    copy: {
      dist: {
        files: [
          {
            expand: true,
            dot: true,
            cwd: '<%= yeoman.app %>/scripts/',
            dest: '<%= yeoman.dist %>/scripts/',
            src: [
              '{,*/}*'
            ]
          }
        ]
      },
      styles: {
        files: [
          {
            expand: true,
            dot: true,
            cwd: '.tmp/styles/',
            dest: '<%= yeoman.dist %>/',
            src: [
              '{,*/}*'
            ]
          }
        ]
      }
    },
    concat: {
      options: {
        separator: ';'
      },
      dist: {
        src: [
          '<%= yeoman.dist %>/scripts/{,*/}*.js'
        ],
        dest: '<%= yeoman.dist %>/<%= yeoman.module %>.js'
      }
    },
    uglify: {
      options: {
        mangle: false,
        banner: '/*! <%= yeoman.module %> v<%= yeoman.version %> */',
        sourceMap: true,
        screwIE8: true
      },
      dist: {
        files: {
          '<%= yeoman.dist %>/<%= yeoman.module %>.min.js': ['<%= yeoman.dist %>/<%= yeoman.module %>.js'],
        }
      }
    },
    usemin: {
      html: ['<%= yeoman.dist %>/{,*/}*.html'],
      css: ['.tmp/styles/{,*/}*.css'],
      options: {
        dirs: ['<%= yeoman.dist %>/images'],
        basedir: ['<%= yeoman.dist %>/images'],
        assetsDirs: ['<%= yeoman.dist %>', '<%= yeoman.dist %>/images']
      }
    },
    concurrent: {
      server: [
        'sass:server',
        'copy:styles'
      ],
      test: [
        'copy:styles'
      ],
      dist: [
        'sass',
        'svgmin'
      ]
    },
    svgmin: {
      options: {
        plugins: {
          removeViewBox: false
        }
      }
    },
    cssmin: {
      options: {
        banner: '/*! <%= yeoman.module %> v<%= yeoman.version %> */',
        sourceMap: true
      },
      target: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.dist %>',
          src: ['*.css', '!*.min.css'],
          dest: '<%= yeoman.dist %>',
          ext: '.min.css'
        }]
      }
    },
    replace: {
      dist: {
        options: {
          patterns: [
            {
              match: /v(\d+)\.(\d+)\.(.+)/g,
              replacement: 'v<%= yeoman.version %>'
            }
          ]
        },
        files: [
          {expand: true, flatten: true, src: ['README.md'], dest: ''}
        ]
      }
    },
    cssUrlEmbed: {
      encodeWithBaseDir: {
        expand: true,
        cwd: '.tmp/styles',
        src: [ '{,*/}*.css' ],
        dest: '.tmp/styles',
        options: {
          baseDir: './src/styles'
        }
      }
    },
    karma: {
      unit: {
        configFile: 'test/karma.conf.js',
        singleRun: false
      },
      build: {
        configFile: 'test/karma.conf.js',
        singleRun: true
      }
    }
  });


  grunt.registerTask('serve', 'Compile then start a connect web server', function (target) {
    if (target === 'dist') {
      return grunt.task.run(['build', 'connect:dist:keepalive']);
    }

    grunt.task.run([
      'clean:server',
      'wiredep',
      'concurrent:server',
      'autoprefixer',
      'connect:livereload',
      'watch'
    ]);
  });

  grunt.registerTask('test', [
    'clean:server',
    'concurrent:test',
    'autoprefixer',
    'connect:test',
    'karma:unit'
  ]);

  grunt.registerTask('build', [
    'karma:build',
    'clean',
    'ngtemplates',
    'replace',
    'copy:dist',
    'concat',
    'uglify:dist',
    'concurrent:dist',
    'autoprefixer',
    // 'cssUrlEmbed',
    'usemin',
    'copy:styles',
    'cssmin',
    'clean:server'
  ]);

  grunt.registerTask('default', [
    'newer:jshint',
    'test',
    'build'
  ]);
};


src/scripts/directives/format.js
'use strict';
(function(module) {
  try {
    module = angular.module('tink.formathelper');
  } catch (e) {
    module = angular.module('tink.formathelper', ['tink.datehelper','tink.safeApply']);
  }
  module.directive('tinkFormatInput', ['dateCalculator', '$window', 'safeApply', function(dateCalculator, $window, safeApply) {
    return {
      restrict: 'EA',
      replace: true,
      priority:99,
      controller:'tinkFormatController',
      controllerAs:'ctrl',
      scope:{
        minDate:'=?',
        maxDate:'=?',
        isDisabled:'=?'
      },
      require:['tinkFormatInput','ngModel','?^form'],
      template: function() {
        var isNative = /(ip(a|o)d|iphone|android)/ig.test($window.navigator.userAgent);
        var isTouch = ('createTouch' in $window.document) && isNative;
        if (isTouch) {
          return '<div><input id="input" class="faux-input" type="date"/><div>';
        } else {
          return '<div><div  id="input" role="textbox" class="faux-input" contenteditable="true" data-ng-bind="placeholder"></div></div>';
        }
      },
      compile: function(template) {
        template.prop('type', 'date');
        return {
          pre: function() {
          },
          post: function(scope, elem, attr, ctrl) {
            // -- check if we are using a touch device  --/
            var isDateSupported = function() {
              var i = document.createElement('input');
              i.setAttribute('type', 'date');
              return i.type !== 'text';
            };

            var isNative = /(ip(a|o)d|iphone|android)/ig.test($window.navigator.userAgent);
            var isTouch = ('createTouch' in $window.document) && isNative && isDateSupported();
        //variable
        var config = {
          format: '00/00/0000',
          placeholder: 'dd/mm/jjjj'
        };
        var dateformat;
        if(isTouch){
          dateformat = 'yyyy-mm-dd';
          config.placeholder = 'yyyy-MM-dd';
        }else{
          dateformat = 'dd/mm/yyyy';
        }

        var element = elem.find('#input');
        var ngControl = ctrl[1];
        var controller = ctrl[0];
        var form = ctrl[2];

        var prefix = '';
        if(angular.isDefined(attr.validName)){
          setTimeout(function(){
            if(form && attr.name && typeof attr.name === 'string' && attr.name !== ''){
            safeApply(scope,function(){
              prefix = attr.validName;
              form.$removeControl(ngControl);
              ngControl.$name = prefix+attr.name;
              form.$addControl(ngControl);
            });
          }
          }, 1);
        }else{
          setTimeout(function(){
            safeApply(scope,function(){
              if(form){
                form.$addControl(ngControl);
              }
            });
          }, 1);
        }

        function validFormat(date,format){
          var dateObject;
          if(angular.isDefined(date) && date !== null){

            if(typeof date === 'string'){
              if(date.length !== 10){ return false; }

              if(!isTouch && !/^(?:(?:31(\/)(?:0?[13578]|1[02]))\1|(?:(?:29|30)(\/|)(?:0?[1,3-9]|1[0-2])\2))(?:(?:1[6-9]|[2-9]\d)?\d{2})$|^(?:29(\/)0?2\3(?:(?:(?:1[6-9]|[2-9]\d)?(?:0[48]|[2468][048]|[13579][26])|(?:(?:16|[2468][048]|[3579][26])00))))$|^(?:0?[1-9]|1\d|2[0-8])(\/|)(?:(?:0?[1-9])|(?:1[0-2]))\4(?:(?:1[6-9]|[2-9]\d)?\d{2})$/.test(date)){return false;}

              dateObject = dateCalculator.getDate(date, format);
            }else if(angular.isDate(date)){
              dateObject = date;
            }else if(typeof date === 'function'){
              return validFormat(date(),format);
            }else {
              return false;
            }

            return dateObject.toString()!=='Invalid Date';
          }
        }

        var isRequired = (function(){
          if(attr.required || attr.require){
            return true;
          }else{
            return false;
          }
        })();
        var noErrorClass = 'hide-error';
        var errorElem = $(elem.find('.faux-input')[0]);

        function checkValidity(value){
          var stringValue;
          if(value != 'Invalid Date' && angular.isDate(value)){
            stringValue = dateCalculator.format(value,dateformat);
          }else{
            stringValue = value;
          }


          safeApply(scope,function(){

            if(angular.isDate(scope.minDate)){
              if(dateCalculator.dateBeforeOther(value,scope.minDate)){
                ngControl.$setValidity('date-min',true);
                errorElem.addClass(noErrorClass);
              }else{
                ngControl.$setValidity('date-min',false);
                errorElem.removeClass(noErrorClass);
              }

            }
            if(angular.isDate(scope.maxDate)){
              if(dateCalculator.dateBeforeOther(scope.maxDate,value)){
                ngControl.$setValidity('date-max',true);
                errorElem.addClass(noErrorClass);
              }else{
                ngControl.$setValidity('date-max',false);
                errorElem.removeClass(noErrorClass);
              }
            }

            if(validFormat(stringValue,dateformat)){
              ngControl.$setValidity('date',true);
              errorElem.addClass(noErrorClass);
              if(isRequired){
                ngControl.$setValidity('required',true);
                errorElem.addClass(noErrorClass);
              }
            }else if(stringValue !== config.placeholder && stringValue !== null){
              ngControl.$setValidity('date',false);
              ngControl.$setValidity('date-min',true);
              ngControl.$setValidity('date-max',true);
              errorElem.removeClass(noErrorClass);
              if(isRequired){
                ngControl.$setValidity('required',true);
              }
            }else{
              ngControl.$setValidity('date',true);
              ngControl.$setValidity('date-min',true);
              ngControl.$setValidity('date-max',true);
              errorElem.addClass(noErrorClass);
              if(isRequired){
                ngControl.$setValidity('required',false);
                errorElem.removeClass(noErrorClass);
              }
            }

          });
        }
        var myWatch = 0;
 //format text going to user (model to view)
        scope.$watch('$parent.'+ attr.ngModel, function(newVal) {
          if(myWatch === 0){
            if(!isTouch || isTouch && newVal !== ''){
              if(newVal != 'Invalid Date' && angular.isDate(newVal)){
                var date = dateCalculator.format(newVal,dateformat);
                controller.setValue(date,null,isTouch,false);
                checkValidity(newVal);
              }else{
                controller.setValue(null,null,isTouch,false);
              }
            }else{
               controller.setValue('',null,isTouch,false);
            }
            checkValidity(newVal);
          }else{
            myWatch = 0;
          }
        }, true);

        controller.init(element,config,form,ngControl);

          //format text from the user (view to model)
          ngControl.$parsers.unshift(function(value) {

              if(isTouch && value === ''){
                value = element.val();
              }

            if(value === null || value === ''){
              return value;
            }else if(typeof value === 'string'){
              controller.setValue(value,null,isTouch);
              return dateCalculator.getDate(value,dateformat);
            }else{
              return null;
            }
          });
          element.unbind('input').unbind('change');

          //on blur update the model.
          element.on('blur', function() {
            safeApply(scope,function(){
              var value;
              var date;
              if(isTouch){
                value = element.val();
              }else{
                var values = $(element).clone().contents().contents().unwrap()[0];
                if(values){
                  value = values.wholeText;
                }
              }
              if(value === config.placeholder || value === undefined){
                checkValidity(value);
                //ngControl.$setViewValue();
              }else{
                date = dateCalculator.getDate(value,dateformat);
                if(date === null){
                  checkValidity(value);
                }else{
                  checkValidity(date);
                }
                controller.setValue(value);
                //fires 2 watches !
                //ngControl.$setViewValue(value);
                //ngControl.$setDirty();
                //ngControl.$render();
              }

              if(!validFormat(date,dateformat)){
                myWatch = 1;
               // controller.setValue(date);
              }
            });
          });
      }
    };

  }
};
}]);
})();